libs/wire-api-federation/src/Wire/API/Federation/API.hs

```diff
@@ -48,12 +49,14 @@ type instance FedApi 'Brig = BrigApi
 
 type instance FedApi 'Cargohold = CargoholdApi
 
-type HasFedEndpoint comp api name = ('Just api ~ LookupEndpoint (FedApi comp) name)
+type HasFedEndpoint comp api name = ('Just api ~ LookupEndpoint (FedApi comp) name, CallsFed comp name)
+
+class CallsFed (comp :: Component) (name :: Symbol)
```

Comment:
I'm yet to look at this thoroughly, but isn't `CallsFed comp name` the same as

```haskell
type CallsFed comp name = forall e. HasFedEndpoint comp e name
```

that is, so long as there exists a federation API endpoint `e` by name `name` in component `comp` (Galley, Brig or Cargohold), `CallsFed comp name` holds?

Reply:
No, `forall` is the universal quantifier ∀. So `forall e. X e` means that the constraint `X` holds for all possible values of `e`. What you are probably thinking about here is `exists` aka `∃`, which haskell doesn't have for whatever reason. You can emulate `exists e. X e` using "impredicative encoding"/CPS/Church encoding as follows:
```haskell
data ExistsX = forall e. ExistsX (X e)
```
This means that `ExistsX` is equipped with a constructor of type `ExistsX :: forall e. X e → ExistsX`, which by the adjunction relating `exists` and `forall`, is equivalent to `(exists e. X e) → ExistsX`.

In any case, I don't think this is very realistic. For one, if I'm not mistaken, existentially quantified constraints are simply not supported. Even if they were, I doubt you would get any sort of reasonable type inference out of them, so you'd still have to specify them manually. Haskell is just not the right language for encoding logic in types.

Reply:
>  so long as there exists a federation API endpoint e by name name in component comp (Galley, Brig or Cargohold), CallsFed comp name holds?

The goal of `CallsFed`  is that it *doesn't have any instances,* ie, it shouldn't be dischargable. This forces all downstream callers to also add the constraint, so that we can guarantee it makes it down to the servant `Server` implementation for the API. I'll add some commentary to that effect.